flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -55,20 +68,33 @@
 			CloseableRegistry closeableRegistry,
 			ExecutorService executorService,
 			Environment env,
-			AsyncExceptionHandler asyncExceptionHandler) {
-		this.checkpointStorage = checkNotNull(checkpointStorage);
+			AsyncExceptionHandler asyncExceptionHandler,
+			boolean sendChannelState) throws IOException {
+		this.checkpointStorage = new CachingCheckpointStorageWorkerView(checkNotNull(checkpointStorage));
 		this.taskName = checkNotNull(taskName);
 		this.closeableRegistry = checkNotNull(closeableRegistry);
 		this.executorService = checkNotNull(executorService);
 		this.env = checkNotNull(env);
 		this.asyncExceptionHandler = checkNotNull(asyncExceptionHandler);
 		this.actionExecutor = actionExecutor;
+		this.channelStateWriter = sendChannelState ? openChannelStateWriter() : ChannelStateWriter.NO_OP;
+		this.closeableRegistry.registerCloseable(this);
+	}
+
+	private ChannelStateWriterImpl openChannelStateWriter() {
+		ChannelStateWriterImpl writer = new ChannelStateWriterImpl(this.checkpointStorage);
+		writer.open();
```

Comment:
This looks a bit suspicious. Do we actually need a dedicated `open` or could we `open` it in the constructor already. Maybe it would allow us to have more final fields?

Reply:
It starts a thread and it doesn't affect fields.